File: ios/Sources/NoiseProtocol/DhDispatch.swift
import Foundation

/// Which key slot to reference in a DH operation.
enum KeyRef: Sendable {
    case s, e, rs, re
}

/// A DH operation descriptor: local key pair + remote public key.
struct DhOp: Equatable, Sendable {
    let local: KeyRef
    let remote: KeyRef
}

/// Complete DH dispatch table — maps (token, role) to key references.
let DH_DISPATCH: [String: [Role: DhOp]] = [
    "ee": [.initiator: DhOp(local: .e, remote: .re),
           .responder: DhOp(local: .e, remote: .re)],
    "es": [.initiator: DhOp(local: .e, remote: .rs),
           .responder: DhOp(local: .s, remote: .re)],
    "se": [.initiator: DhOp(local: .s, remote: .re),
           .responder: DhOp(local: .e, remote: .rs)],
    "ss": [.initiator: DhOp(local: .s, remote: .rs),
           .responder: DhOp(local: .s, remote: .rs)],
]

File: android/src/main/kotlin/noise/protocol/DhDispatch.kt
package noise.protocol

/** Which key slot to reference in a DH operation. */
enum class KeyRef { S, E, RS, RE }

/** A DH operation descriptor: local key pair + remote public key. */
data class DhOp(val local: KeyRef, val remote: KeyRef)

/**
 * Complete DH dispatch table — maps (token, role) to key references.
 *
 * Replaces the role-conditional `when`/`if` branches in processDHToken
 * with a pure data lookup. Exhaustive and independently testable.
 */
val DH_DISPATCH: Map<Pair<String, Role>, DhOp> = mapOf(
    ("ee" to Role.INITIATOR) to DhOp(KeyRef.E, KeyRef.RE),
    ("ee" to Role.RESPONDER) to DhOp(KeyRef.E, KeyRef.RE),
    ("es" to Role.INITIATOR) to DhOp(KeyRef.E, KeyRef.RS),
    ("es" to Role.RESPONDER) to DhOp(KeyRef.S, KeyRef.RE),
    ("se" to Role.INITIATOR) to DhOp(KeyRef.S, KeyRef.RE),
    ("se" to Role.RESPONDER) to DhOp(KeyRef.E, KeyRef.RS),
    ("ss" to Role.INITIATOR) to DhOp(KeyRef.S, KeyRef.RS),
    ("ss" to Role.RESPONDER) to DhOp(KeyRef.S, KeyRef.RS),
)
